raphtory/src/db/api/state/ops/node.rs

```diff
@@ -0,0 +1,146 @@
+use crate::{
+    db::api::{storage::graph::storage_ops::GraphStorage, view::internal::CoreGraphOps},
+    prelude::GraphViewOps,
+};
+use raphtory_api::core::{
+    entities::{GID, VID},
+    storage::arc_str::ArcStr,
+    Direction,
+};
+use std::{ops::Deref, sync::Arc};
+
+pub trait NodeOp: Send + Sync {
+    type Output: Clone + Send + Sync;
+    fn apply(&self, storage: &GraphStorage, node: VID) -> Self::Output;
+
+    fn map<V: Clone + Send + Sync>(self, map: fn(Self::Output) -> V) -> Map<Self, V>
+    where
+        Self: Sized,
+    {
+        Map { op: self, map }
+    }
+}
+
+// Cannot use OneHopFilter because there is no way to specify the bound on Output
+pub trait NodeOpFilter<'graph>: NodeOp + 'graph {
+    type Graph: GraphViewOps<'graph>;
+    type Filtered<G: GraphViewOps<'graph>>: NodeOp<Output = Self::Output>
+        + NodeOpFilter<'graph, Graph = G>
+        + 'graph;
+
+    fn graph(&self) -> &Self::Graph;
+
+    fn filtered<G: GraphViewOps<'graph>>(&self, graph: G) -> Self::Filtered<G>;
+}
+
+#[derive(Debug, Clone, Copy)]
```

Comment:
I wonder if many of these could just be special cases of a more generic 
``` struct NodeOpFn<O, F: Fn(&GraphStorage, VID) -> O>(F)```
and could be just made from constructors like 
``` impl NodeOpFn {
fn id()-> ...
```

Reply:
If you do this you can no longer name the type which causes a bunch of problems, though maybe we can get away with `fn`